packages/vite/src/node/optimizer/index.ts

```diff
@@ -761,7 +761,12 @@ async function prepareEsbuildOptimizerRun(
     ),
   }
 
-  const platform = environment.config.consumer === 'client' ? 'browser' : 'node'
+  const platform =
+    environment.config.optimizeDeps.esbuildOptions?.platform ??
+    // We generally don't want to use platform 'neutral', as esbuild has custom handling
+    // when the platform is 'node' or 'browser' that can't be emulated by using mainFields
+    // and conditions
+    (environment.config.consumer === 'client' ? 'browser' : 'node')
```

Comment:
```suggestion
  const platform =
    optimizeDeps.esbuildOptions?.platform ??
    // We generally don't want to use platform 'neutral', as esbuild has custom handling
    // when the platform is 'node' or 'browser' that can't be emulated by using mainFields
    // and conditions
    (environment.config.consumer === 'client' ||
    environment.config.ssr.target === 'webworker'
      ? 'browser'
      : 'node')
```
I also found that I forgot to add a backward compat for `ssr.target === 'webworker'`.